File: solutions/3904-minimum-adjacent-swaps-to-alternate-parity/minimum-adjacent-swaps-to-alternate-parity.cpp
class Solution {
public:
    int minSwaps(std::vector<int>& nums) {
        std::vector<int> ones;
        int n = nums.size();

        for (int i = 0; i < n; ++i) {
            if (nums[i] % 2 == 1) {
                ones.push_back(i);}
        }

        int m = 2 * ones.size();  

        auto countSwaps = [&](int start) {
            int swaps = 0;
            for (int i = 0; i < ones.size(); ++i) {
                swaps += std::abs(ones[i] - (start + 2 * i));}
            return swaps; };

        if (m == n) {
            return std::min(countSwaps(0), countSwaps(1));}
        if (m == n + 1) {
            return countSwaps(0);}
        if (m == n - 1) {
            return countSwaps(1);}
        return -1;}
};

File: solutions/3904-minimum-adjacent-swaps-to-alternate-parity/minimum-adjacent-swaps-to-alternate-parity.java
class Solution {
    public int minSwaps(int[] nums) {
        List<Integer> ones = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n; ++i) {
            if (nums[i] % 2 == 1) {
                ones.add(i);}}

        int m = 2 * ones.size();

        if (m == n) {
            return Math.min(countSwaps(ones, 0, n), countSwaps(ones, 1, n));}
        if (m == n + 1) {
            return countSwaps(ones, 0, n);}
        if (m == n - 1) {
            return countSwaps(ones, 1, n); }
        return -1; }

    private int countSwaps(List<Integer> ones, int start, int n) {
        int swaps = 0;
        
        for (int i = 0; i < ones.size(); ++i) {
            swaps += Math.abs(ones.get(i) - (start + 2 * i));}
        return swaps;}
}